render.go

```diff
@@ -77,37 +80,41 @@ func (r *Renderer) DrawShapedRunAt(run shaping.Output, img draw.Image, startX, s
 		r.PixScale = 1
 	}
 	scale := r.FontSize * r.PixScale / float32(run.Face.Upem())
+	r.fillerScale = scale
 
 	b := img.Bounds()
 	scanner := rasterx.NewScannerGV(b.Dx(), b.Dy(), img, b)
 	f := rasterx.NewFiller(b.Dx(), b.Dy(), scanner)
+	r.filler = f
 	f.SetColor(r.Color)
 	x := float32(startX)
 	y := float32(startY)
 	for _, g := range run.Glyphs {
 		x -= fixed266ToFloat(g.XOffset) * r.PixScale
-		outline, _ := run.Face.GlyphData(g.GlyphID).(api.GlyphOutline)
-
-		for _, s := range outline.Segments {
-			switch s.Op {
-			case api.SegmentOpMoveTo:
-				f.Start(fixed.Point26_6{X: floatToFixed266(s.Args[0].X*scale + x), Y: floatToFixed266(-s.Args[0].Y*scale + y)})
-			case api.SegmentOpLineTo:
-				f.Line(fixed.Point26_6{X: floatToFixed266(s.Args[0].X*scale + x), Y: floatToFixed266(-s.Args[0].Y*scale + y)})
-			case api.SegmentOpQuadTo:
-				f.QuadBezier(fixed.Point26_6{X: floatToFixed266(s.Args[0].X*scale + x), Y: floatToFixed266(-s.Args[0].Y*scale + y)},
-					fixed.Point26_6{X: floatToFixed266(s.Args[1].X*scale + x), Y: floatToFixed266(-s.Args[1].Y*scale + y)})
-			case api.SegmentOpCubeTo:
-				f.CubeBezier(fixed.Point26_6{X: floatToFixed266(s.Args[0].X*scale + x), Y: floatToFixed266(-s.Args[0].Y*scale + y)},
-					fixed.Point26_6{X: floatToFixed266(s.Args[1].X*scale + x), Y: floatToFixed266(-s.Args[1].Y*scale + y)},
-					fixed.Point26_6{X: floatToFixed266(s.Args[2].X*scale + x), Y: floatToFixed266(-s.Args[2].Y*scale + y)})
+		data := run.Face.GlyphData(g.GlyphID)
+		switch format := data.(type) {
+		case api.GlyphOutline:
+			adv := r.drawOutline(g, format, f, scale, x, y)
```

Comment:
I think you could avoid computing the advance again: using the glyph XAdvance should be more reliable